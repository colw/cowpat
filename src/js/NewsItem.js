import React from "react";
import moment from 'moment';

const LinkExt = (props) => (
  <a {...props} target="_blank" rel="noopener"/>
)

export default class NewsItem extends React.Component {

  state = {formattedTimeSince: moment(this.props.info.fetchDate).fromNow()};

  componentDidMount () {
    this.setInterval(this.updateTime.bind(this), 30000);
  }

  updateTime () {
    this.setState({formattedTimeSince: moment(this.props.info.fetchDate).fromNow()});
  }

  getBaseURL (url) {
    if (url.slice(0,4) != 'http') {
      url = 'http://' + url;
    }
    var a = document.createElement('a');
    a.href = url;
    return a.hostname.replace(/^www./, '');
  }

  componentWillMount () {
    this.intervals = [];
  }

  setInterval () {
    this.intervals.push(setInterval.apply(null, arguments));
  }

  componentWillUnmount () {
    this.intervals.map(clearInterval);
  }

	render () {
    var hosturl = this.getBaseURL(this.props.info.metalink);
		return (
			<div className="newsItem">
				<LinkExt href={this.props.info.link}>
					<div className="headTitle">
            {this.props.info.title}
					</div>
        </LinkExt>
        <div className="subTitle"><LinkExt href={this.props.info.sitelink}>{hosturl}</LinkExt></div>
			</div>
		);
	}
};