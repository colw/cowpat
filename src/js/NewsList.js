import React from "react";
import moment from 'moment';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group'

import NewsItem from "./NewsItem";
import NewsCow from "./NewsCow";

// const NewsItem = ({info}) => (
//   <div className="newsItem">
//     {info.title} <span className="subItem">{info.sitelink}</span>
//     <div style={{position: 'absolute', bottom: 0, right: 0}}>
//       <img src={`/dist/icons/${info.sitehost}.jpg`} />
//     </div>
//   </div>
// )

export default class NewsList extends React.Component {
  render () {

    if (this.props.loading) {
      return <div>Loading</div>
    }


    if (this.props.newsItems.length === 0) {
      return (
        <div id="emptyList">
        </div>
      );
    } else {
      var makeList = x => <li key={x.guid} className="newsItemWrapper"><NewsItem info={x}/></li>;

      return (
          <ReactCSSTransitionGroup component="ul" className="newsWrapper"
            transitionName="example"
            transitionAppear={true}
            transitionAppearTimeout={500}            
            transitionEnterTimeout={500}
            transitionLeaveTimeout={300}>
            { this.props.newsItems.map(makeList) }
          </ReactCSSTransitionGroup>
  		);
    }
	}
};