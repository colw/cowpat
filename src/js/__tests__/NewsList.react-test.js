// NewsList.react-test.js
import React from 'react';
import NewsItem from '../NewsItem';
import renderer from 'react-test-renderer';
import NewsList from '../NewsList';

const data = {
	"newsItems": [{
		"author": "Associated Press",
		"guid": "https://www.washingtonpost.com/world/asia_pacific/afghan-official-7-killed-in-a-suicide-attack/2017/02/11/e3879406-f053-11e6-a100-fdaaf400369a_story.html",
		"link": "https://www.washingtonpost.com/world/asia_pacific/afghan-official-7-killed-in-a-suicide-attack/2017/02/11/e3879406-f053-11e6-a100-fdaaf400369a_story.html",
		"title": "Afghan official: 7 killed in a suicide attack",
		"sitehost": "www.washingtonpost.com",
		"itemID": "01c6f30bd8afd1d464662450eaa3e2e0",
		"metalink": "http://www.washingtonpost.com/pb/world/",
		"tags": "[\"afghan official\",\"killed\",\"suicide attack\"]",
		"metatitle": "World",
		"date": "Thu Jan 01 1970 00:00:00 GMT+0000 (UTC)",
		"sitelink": "http://www.washingtonpost.com",
		"description": "An Afghan official says at least seven people were killed when a suicide bomber attacked Afghan soldiers in southern Helmand province."
	}],
	"filterTags": "[\"afghan official\",\"killed\",\"suicide attack\"]"
};

test('NewsList Stub', () => {
  const component = renderer.create(
    <NewsList key={1} loading={true} newsItems={data.newsItems} filterText={''} filterTags={data.filterTags}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

