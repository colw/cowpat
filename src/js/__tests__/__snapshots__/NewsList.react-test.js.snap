exports[`test NewsList Stub 1`] = `
<div>
  Loading
</div>
`;
